working/extension_structs/overview.md

```diff
@@ -0,0 +1,815 @@
+
+### Const structs
+
+TODO(leafp): This should work, write out the details.
+
+## Extension Structs
+
+Extension structs are restrictions of the core struct feature, designed to
+support wrapper-less views on an object.  This section describes how extension
+structs look and behave, largely by example.  In many cases, alternatives or
+extensions to the core proposal are described in line.
+
+## Extension structs by example
+
+Extension structs are targeted at relatively niche uses where you wish to define
+a set of statically dispatched methods layered on top of an underlying
+representation, without introducing a wrapper object.  A canonical use case
+driving this design is to be able define a Dart typed interface for methods on a
+Javascript object, providing a wrapperless interoperation capability.
+
+
+```dart
+extension struct Window(JSObject o) {
+  // Signatures provided here for methods to be delegated to the underlying
+  // JSObject
+  external bool get closed;
+  // etc
+}
+
+void test(Window w) {
+  // Window methods can be called using Dart syntax
+  if (w.closed) {...}
+
+  // Windows are represented as the underlying object
+  assert(w is JSObject);
+
+  // Windows can be cast to the underlying object
+  JSObject o = w as JSObject;
+
+  // Extension struct types are reified as the underlying representation type.
+  List<Window> l = [w];
+  assert(l is List<JSObject>);
+}
+```
+
+Extension structs are also useful for providing a lightweight facade over an
+existing type.
+
+```dart
+// Natural numbers
+extension struct Nat(int _x) {
+  Nat(int x) : assert(x >= 0), _x = x;
+  Nat.zero() : _x = 0;
+
+  Nat get succ => Nat(_x+1);
+  Nat plus(Nat other) => Nat._x + other._x;
+
+  // Override the underlying isNegative operation (incorrectly)
+  bool get isNegative => true;
+}
+
+void test() {
+  var n1 = Nat(3);
+  var n2 = Nat.zero();
+  assert(n2.succ.succ.succ == n2);
+
+  //The underlying representation is still as an int
+  assert(n1 is int);
+
+  // The static type is used to dispatch the method calls
+  assert(n1.isNegative);
+
+  // If the static type is lost, dispatch goes to the underlying object.
+  dynamic d = n1;
+  assert(!d.isNegative);
+}
+```
+
+Extension structs may delegate members to the underlying field.  They may also
+implement interfaces, but only if the underlying field implements the interface.
+
+```dart
+// Natural numbers
+extension struct Nat(int _x) implements Comparable<num> {
+  // Constructors etc as above
+
+
+  bool get isEven;  // Abstract definition delegates to _x.isEven
+}
+
+void test() {
+  var n = Nat(2);
+  // Same as 2.isEven
+  assert(n.isEven);
+  // Comparable interface allows access to the Comparable methods on int
+  assert(n.compareTo(2) == 0);
+
+  // Since Nat implements Comparable<num>, it may be assigned to it
+  Comparable<num> c = n;
+  // The representation is still as an integer
+  assert(c is int);
+}
+```
+
+Extension structs have no inheritance.
```

Comment:
If the worry with extension is that we are adding fields, this is one restriction that would work for us:

structs can only extend other structs that have the exact same set of fields and default values. In the case of JS interop, it's the same JSObject after all.

Reply:
I think there's something that can hang together. I'll try to add a section proposing how this could work on Monday.  It would be useful to know how much is actually needed though.  Do you need mixins?  Do you need interfaces?  Do you need concrete superclasses, or are abstract superclasses enough?

Reply:
I believe the main requirement we have is concrete superclasses. I don't believe we need mixins or interfaces, but @srujzs and @rileyporter may have more context here.

A common example is something like this:
```dart
static struct Node(JSObject o) {
  external get childNodes;
}

static struct Element(JSObject o) extend Node(o) {
  external get name;
}
```

(even though they have no body as declared, I'd default to reason about `external` as a concrete members, since we also have scenarios where we need to add method bodies)

Another common example is that we want to have some features layered:
'default.dart':
```dart
static struct Element(JSObject o) {
  external addEventListener;
}
```

'compat.dart':
```dart
import 'default.dart' as default;
static struct Element(JSObject o) extends default.Element(o) {
  Stream<Event> onClick => ....
}
```

`compat.dart` provides some backwards compatibility while we migrate from one API to another by using additional extensions.

This could be done with double wrapping too, but because the first wrapper has no interface, it makes it harder to "reexport" it's API.

Reply:
The main problem that I have with concrete superclasses is actually not with extension structs, but with regular structs.  I was trying to avoid structs having virtual fields.  I can probably still cobble something together, I'll take a crack next week.

Reply:
Thinking more about my last sentence above - I think we could go a long way with double-wrappers if we can find a way to forward the wrapper static interface:

```dart
static struct A(int x) {
  bool isFour => x == 4;
}
static struct B(A a) implements A {
  bool isFive => a == 5; // this implicitly means a.x == 5
}

main() {
  B b = B(A(generateInt()));
  print(b.isFour); // valid
```

I don't like the word `implements` here, since it seems to imply that `A` has an interface. But the notion that we can expose parts of the underlying wrapped API statically. But maybe keywords like `show` (if I recall I think Erik once proposed that) could work

Reply:
I think this is what I would make `extends` do, FWIW.

Reply:
We don't really support mixins today for interop and I think I've maybe seen a total of 2 bugs that tried to use a mixin with interop. They may work today in some cases, but I do not see a strong need for them to be included in the future. I talk about interfaces in the other comment chain.

Reply:
Yes, I think that's my initial take on adding `extends`.  We say that `extends` means that all methods on the superview are available on the subview, but you can't override them.  @sigmundch @srujzs do we need to allow override?  We can do so, but it has the obvious surprising behavior when you assign from a subview type to a superview type and then access the method (and get the superview method rather than the subview one).

@sigmundch @srujzs do you feel that we need to allow cross-library extension, or can we keep views "open in the library, sealed externally"?

Reply:
> We can do so, but it has the obvious surprising behavior when you assign from a subview type to a superview type and then access the method (and get the superview method rather than the subview one)

Maybe I'm coming at this weird, but this doesn't seem surprising to me. :) You "view" it differently, so you're going to get a different behavior/method. I personally would want this behavior (but maybe the deviation from classes is indeed confusing).

> do we need to allow override?

In the case where all the methods are external anyways, no, because JS will do the appropriate dynamic dispatch. Obviously, it gets a little more complicated in the cases where you want to do define non-external methods for whatever reason. One instance I can see in `dart:html` is different behavior for `setInnerHtml` for sanitization reasons (see `TemplateElement`).

Reply:
> Maybe I'm coming at this weird, but this doesn't seem surprising to me. :)

It's not surprising to me either, but my goal with this proposal was that for *consumers* of the API, they should as much as possible be able to guess what things would do based solely on knowing what classes and objects do.  That is, I want developers to not really have to know what views are to be able to successfully use an API written using them.

Reply:
Re: need for overrides. I believe we do, but to a lower degree. @rileyporter has examples from the html prototype that illustrate the need for us to shadow a property with a new implementation in a new static view. In our case it is because we want to use a static view to provide backward compatibility to an old API that happens to use the same property name as a new API.

Re: cross-library extension. We believe we need it too. Our approach for exposing html APIs will require a layering approach with an initial set of views defined in one library, and a separate library defining additional static views that extend the base ones.

A couple thoughts:
 * It may be worth revisiting the terminology. The `extends` word (and similarly `implements`) gives the illusion that we have virtual dispatch, and hence the possibility of `overrides`. I personally don't see this part of the feature as an "override", but as a shadowing (more below). So maybe using a different terminology may help our end users as well?
 *  I see a static view as a way of encapsulating access to an underlying object, so I'd like a static view to have a choice of what that static API will be. For example, if I don't want to expose anything from the underlying object I should be allowed to do so. With copy/paste we can achieve anything, but I'd hope we can make the language syntax support the common patterns.

Here are some of the scenarios I'd like to support:

1. Create opaque views:

```dart
// A1 is an opaque view of `int`
// the original methods from `int` are not exposed.
static struct A1(int x) {
  bool isFour => x == 4;
}
```

2. Create opaque views with methods that shadow what exist in the underlying `on` type.
```dart
// A2 can expose APIs with a name that conflict
// with existing APIs in the `on` type, no problem:
static struct A2(int x) {
  bool isEven => throw "WAT";
}
// Yes (A2(4) as int).isEven is true
```

3. Create transparent views, that expose methods from the underlying `on` type.
```dart
// A3 has the same methods as `int`.
// We could say that we always expose all methods on `Object` by default.
static struct A3(int x) show int.* {
}

// The "show" is syntactic sugar for writing:
static struct A3(int x) {
  bool isEven => x.isEven;
  bool isOdd => x.isOdd;
  ... // all other int methods here
}
```

4. Create static views exposing some underlying APIs and shadowing others:
```dart
// A4 hides `isEven` from int, and as result it's OK to shadow it:
static struct A4(int x) show int.* hide int.isEven {
 bool isEven => throw "what";
}
```

5. Detect when you are inadvertently shadowing an API:
```dart
static struct A5(int x) show int.* { // oops this shows isEven
 bool isEven => throw "what"; // ERROR: can't shadow it too!
 // I could be convinced however of some shadowing/scoping rules,  but no conflict
 // like we do with our import syntax seems like a good starting point
}
```

6. Derive static views from other static views:
```dart
// If we use composition as the syntax for deriving subviews, it may look as such:
// This exposes the static members in A1, but still hides anything from `int` (since int was opaquely hidden in A1)
static struct B6(A1 a) show A1.* {
  bool isNotEven => !(a as int).isEven; // cast needed because A1 was opaque and didn't expose isEven
  // but A1.isFour is available on B6 variables
}
```

Thoughts?

Reply:
This proposal supports all of your examples, 1 through 6, though in some cases with slightly more verbose syntax.

> It may be worth revisiting the terminology. The extends word (and similarly implements) gives the illusion that we have virtual dispatch, and hence the possibility of overrides. I personally don't see this part of the feature as an "override", but as a shadowing (more below). So maybe using a different terminology may help our end users as well?

So, this proposal is very specifically trying to avoid doing what you are proposing here.  It's almost the heart of the design principle here.  I don't think we should be designing views to be the way that Dart users design large amounts of new Dart code.  I think they are filling a fairly niche area.  As such, I would like to minimize the cognitive load on *consumers* of views.  If a user sees `static struct B6(A1 a) show A1.*`, they can... probably guess what it does?  But compare that to `view class B6(int x) extends A1` .  Sure, they need to learn what primary constructors are (but they learn that for data classes too).  But other than that, all of their existing intuitions carry over.  If Foo extends Bar, then Foo is a subtype of Bar, so B6 is a subtype of A1.  check.  If Foo extends Bar, then Foo has all of the methods of Bar.  check.  Now, of course, there are differences that will show through.  But the idea behind this design is to see how much we can leverage existing behavior to make the view feature be as small of a cognitive load on Dart users as possible.  Now, maybe that doesn't work out, in which case maybe we take some pieces of this and some pieces of the other proposal (e.g. your hybrid primary constructor + show hide examples above)?

> Re: need for overrides. I believe we do,

Ok.

> Re: cross-library extension. We believe we need it too. 

Ok.   My main concern with this is harmonizing it with regular structs, but that's tractable. Maybe this pushes in the direction of splitting apart structs and extension structs into data classes and view classes, so there is less of an imperative to make them consistent on things like this.

It's also not clear that making regular structs sealed is going to fly.  :)

Reply:
> ... So, this proposal is very specifically trying to avoid doing what you are proposing here. It's almost the heart of the design principle here. 

Totally makes sense - I was trying to balance that with the possible cognitive load of having overlapping concepts have different meanings 🙂 

As I think through these extension examples and how do we craft these features in terms of other simpler ones, one more thought comes to mind. Any thoughts about having a general notion of statically dispatched non-virtual methods in classes? We have that with extension methods, but what if they we allow them generally?

I am visualizing the complement of "virtual" in other languages:
```dart
class A1 {
  int m1() => 1;
  nonvirtual int m2() => 2;
}
```
We could consider extends also exposing nonvirtual methods (just like extension methods do):
```dart
class A2 extends A1 {
   // inherits both m1 and m2, but m2 is dispatched statically
}
```

And we could consider allowing overrides or not:

```dart
class A2 extends A1 {
   nonvirtual int m2() => 3; // TBD if we allow it
}

// If we do, nonvirtual methods are dispatched statically:
// (A2() as A1).m2() returns 2
```

I know we are trying to keep the new feature as small as possible, but can't help to connect how all these building blocks are themselves useful features independently. One reason that makes this specific feature attractive is that now static views are even less special. Extension is the same as in regular classes, nothing new here. That is, `static views` then are regular classes that only have a single field, only declare and inherit nonvirtual methods, don't expose an interface (seals them to some extend), etc.

Look forward to your thoughs!

Reply:
It's an interesting idea.  We've discussed non virtual methods before.  C# has what you describe, basically: methods which are dispatched based on type (so even if you "override" them in a subclass, you get the superclass method any time you access via the superclass type).  I agree that it would help a bit with the consistency story here, though possibly only be extending the cognitive load universally... :)

One tricky corner for us would be generics.  For views/extension types, we definitely have the semantics that generic parameters get filled in based on the static type.  We could do the same for non-virtual methods on classes, but it's not a forced move, and in some ways feels off.  Example:

```
class A<T> {
  nonvirtual void show() { print(T); };
}

void main() {
  A<num> a = A<int>();
  a.show();
}
```

Does this print `num`, or `int`?  The extension method semantics give you `num`.  Essentially, we're just giving you a short syntax for defining an extension method on `A`.  But if I'm just thinking of these as ordinary (but non-virtual) methods, I'd kind of expect to see `int`.

Reply:
I filed #2368, #2369, and #2370 to discuss this.  Let's continue the discussion there.  Feel free to copy over anything you feel important to capture.